srcjs/utils.js

```diff
@@ -246,3 +246,230 @@ function mapValues(obj, f) {
+// When the return value of a call to `this.recur` is returned by a method, the
+// arguments that were supplied to `this.recur` are used to call the
+// multimethod.
+//
+// # Examples
+//
+// Handling events:
+//
+//    var handleEvent = multimethod()
+//     .dispatch(e => [e.target.tagName.toLowerCase(), e.type])
+//     .when(["h1", "click"], e => "you clicked on an h1")
+//     .when(["p", "mouseover"], e => "you moused over a p"})
+//     .else(e => {
+//       let tag = e.target.tagName.toLowerCase();
+//       return `you did ${e.type} to an ${tag}`;
+//     });
+//
+//    $(document).on("click mouseover mouseup mousedown", e => console.log(handle(e)))
+//
+// Self-calls:
+//
+//    var demoSelfCall = multimethod()
+//     .when(0, function(n) {
+//       this(1);
+//     })
+//     .when(1, function(n) {
+//       doSomething(this);
+//     })
+//     .when(2, _ => console.log("tada"));
+//
+// Using (abusing?) the test function:
+//
+//    var fizzBuzz = multimethod()
+//     .test((x, divs) => divs.map(d => x % d === 0).every(Boolean))
+//     .when([3, 5], x => "FizzBuzz")
+//     .when([3], x => "Fizz")
+//     .when([5], x => "Buzz")
+//     .else(x => x);
+//
+//    for(let i = 0; i <= 100; i++) console.log(fizzBuzz(i));
+//
+// Getting carried away with tail recursion:
+//
+//    var factorial = multimethod()
+//     .when(0, () => 1)
+//     .when(1, (_, prod = 1) => prod)
+//     .else(function(n, prod = 1) {
+//       return this.recur(n-1, n*prod);
+//     });
+//
+//    var fibonacci = multimethod()
+//     .when(0, (_, a = 0) => a)
+//     .else(function(n, a = 0, b = 1) {
+//       return this.recur(n-1, b, a+b);
+//     });
+function multimethod(dispatch = (firstArg) => firstArg,
+                     test = equal,
+                     defaultMethod = null,
+                     methods = []) {
+
+  var trampolining = false;
+
+  function Sentinel (args) { this.args = args; }
+
+  function trampoline(f) {
+    return (...args) => {
+      trampolining = true;
+      var ret = f.apply(invoke, args);
+      while (ret instanceof Sentinel)
+        ret = f.apply(invoke, ret.args);
+      trampolining = false;
+      return ret;
+    };
+  }
+
+  let invoke = trampoline((...args) => {
+    var dispatchVal = dispatch.apply(null, args);
+    for (let i = 0; i < methods.length; i++) {
+      let [methodVal, methodFn] = methods[i];
+      if (test(dispatchVal, methodVal)) {
+        return methodFn.apply(invoke, args);
+      }
+    }
+    if (defaultMethod) {
+      return defaultMethod.apply(invoke, args);
+    } else {
+      throw new Error(`No method for dispatch value ${dispatchVal}`);
+    }
+  });
+
+  invoke.recur = (...args) => {
+    if (!trampolining) throw new Error("recur can only be called inside a method");
+    return new Sentinel(args);
+  };
+
+  invoke.dispatch = (newDispatch = (firstArg) => firstArg) => {
+    dispatch = newDispatch;
+    return invoke;
+  };
+
+  invoke.test = (newTest = _equal) => {
```

Comment:
Should this be `equal` instead of `_equal`, for consistency? Also, it seems to me that a default value isn't really necessary here, or for `invoke.dispatch`.